lib/Transforms/ForwardInsertToExtract/ForwardInsertToExtract.cpp

```diff
@@ -0,0 +1,130 @@
+#include "lib/Transforms/ForwardInsertToExtract/ForwardInsertToExtract.h"
+
+#include <utility>
+
+#include "llvm/include/llvm/ADT/TypeSwitch.h"  // from @llvm-project
+#include "llvm/include/llvm/Support/Debug.h"   // from @llvm-project
+#include "mlir/include/mlir/Dialect/Affine/IR/AffineOps.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Affine/Utils.h"      // from @llvm-project
+#include "mlir/include/mlir/Dialect/MemRef/IR/MemRef.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/Tensor/IR/Tensor.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/MLIRContext.h"            // from @llvm-project
+#include "mlir/include/mlir/IR/PatternMatch.h"           // from @llvm-project
+#include "mlir/include/mlir/IR/Value.h"                  // from @llvm-project
+#include "mlir/include/mlir/IR/ValueRange.h"             // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"              // from @llvm-project
+#include "mlir/include/mlir/Support/LogicalResult.h"     // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+
+#define DEBUG_TYPE "forward-insert-to-extract"
+
+namespace mlir {
+namespace heir {
+
+#define GEN_PASS_DEF_FORWARDINSERTTOEXTRACT
+#include "lib/Transforms/ForwardInsertToExtract/ForwardInsertToExtract.h.inc"
+
+bool ForwardSingleInsertToExtract::isForwardableOp(
+    Operation *potentialInsert, tensor::ExtractOp &extractOp) const {
+  if (!dominanceInfo.properlyDominates(potentialInsert,
+                                       extractOp.getOperation())) {
+    LLVM_DEBUG(llvm::dbgs() << "insert op does not dominate extract op\n");
+    return false;
+  }
+
+  if (extractOp->getBlock() != potentialInsert->getBlock()) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "insert and extract op are not in the same block\n");
+    return false;
+  }
+
+  return llvm::TypeSwitch<Operation &, bool>(*potentialInsert)
+      .Case<tensor::InsertOp>([&](auto insertOp) {
+        ValueRange insertIndices = insertOp.getIndices();
+        ValueRange extractIndices = extractOp.getIndices();
+        if (insertIndices != extractIndices) {
+          LLVM_DEBUG(llvm::dbgs()
+                     << "insert and extract op do not have matching indices\n");
+          return false;
+        }
+
+        // Naively scan through the operations between the two ops and check if
+        // anything prevents forwarding.
+        for (auto currentNode = insertOp->getNextNode();
+             currentNode != extractOp.getOperation();
+             currentNode = currentNode->getNextNode()) {
+          if (currentNode->getNumRegions() > 0) {
+            LLVM_DEBUG(llvm::dbgs() << "an op with control flow is between the "
+                                       "insert and extract op\n");
+            return false;
+          }
+
+          if (auto op = dyn_cast<tensor::InsertOp>(currentNode)) {
```

Comment:
I was going to say that `tensor.insert_slice` and `tensor.parallel_insert_slice` should also be guarded here (https://mlir.llvm.org/docs/Dialects/TensorOps/#tensorinsert_slice-tensorinsertsliceop), but I actually realize that since each of these operations returns a *new* tensor, that even this branch of code would never hit. For example, 

```
%inserted = tensor.insert %c0 into %arg[%c0_index]
%inserted_1 = tensor.insert %c1 into %arg[%c0_index]
%extract = tensor.extract %inserted[%c0_index]
```

would still mean that we can forward the %c0 scalar value to replace %extract.

Reply:
> %inserted = tensor.insert %c0 into %arg[%c0_index]
> %inserted_1 = tensor.insert %c1 into %arg[%c0_index]
> %extract = tensor.extract %inserted[%c0_index]

used this as 'Forwarding with an insert between' test